File: 백준/위상정렬/2252_줄세우기.py
# 문제 : https://www.acmicpc.net/problem/2252

import sys
from collections import deque
input = sys.stdin.readline

n, m = map(int, input().split())
path = [[] for _ in range(n+1)]
degree = [0] * (n+1)
answer = []

for _ in range(m):
    a, b = map(int, input().split())
    path[a].append(b)
    degree[b] += 1

queue = deque()
for i in range(1, n+1):
    if degree[i] == 0:
        queue.append(i)

while queue:
    num = queue.popleft()
    answer.append(str(num))

    for i in path[num]:
        degree[i] -= 1
        if degree[i] == 0:
            queue.append(i)
print(' '.join(answer))

File: kotlin/백준/위상정렬/2252_줄 세우기.kt
// 문제 : https://www.acmicpc.net/problem/2252

import java.io.BufferedReader
import java.io.InputStreamReader

val br = BufferedReader(InputStreamReader(System.`in`))
fun inputInt() = br.readLine().toInt()
fun inputIntList() = br.readLine().split(" ").map { it.toInt() }

val degree = IntArray(32001)
val graph = Array(32001) { mutableListOf<Int>() }

fun main() {
    val (n, m) = inputIntList()
    repeat(m) {
        val (a, b) = inputIntList()

        degree[a] += 1
        graph[b].add(a)
    }

    val queue = ArrayDeque<Int>()
    val answer = mutableListOf<Int>()

    for (i in 1 .. n) {
        if (degree[i] == 0) {
            queue.add(i)
        }
    }

    while (queue.isNotEmpty()) {
        val current = queue.removeFirst()
        answer.add(current)

        for (next in graph[current]) {
            degree[next] -= 1
            if (degree[next] == 0) {
                queue.add(next)
            }
        }
    }

    println(answer.reversed().joinToString(" "))
}